aioxmpp/ibb/service.py

```diff
@@ -0,0 +1,507 @@
+
+        if low > high:
+            low = high
+
+        self._output_buffer_limit_low = low
+        self._output_buffer_limit_high = high
+
+    def get_write_buffer_limits(self):
+        return self._output_buffer_limit_low, self._output_buffer_limit_high
+
+    def get_write_buffer_size(self):
+        return len(self._write_buffer)
+
+    def set_protocol(self, proto):
+        self._protocol = proto
+        proto.connection_made(self)
+
+    def get_protocol(self):
+        return self._protocol
+
+    def pause_reading(self):
+        self._reading_paused = True
+
+    def resume_reading(self):
+        self._reading_paused = False
+        if self._protocol is not None:
+            self._protocol.data_received(b"".join(self._input_buffer))
+        self._input_buffer.clear()
+
+    def is_closing(self):
+        return self._closing or self._closed
+
+    def get_extra_info(self, key, default=None):
+        return {
+            "block_size": self._block_size,
+            "peer_jid": self._peer_jid,
+            "stanza_type": self._stanza_type,
+            "sid": self._sid,
+        }.get(key, default)
+
+    @asyncio.coroutine
+    def _write_task_main(self):
+        while True:
+            yield from self._can_write.wait()
+
+            if self._write_buffer:
+                data = self._write_buffer[:self._block_size]
+                self._write_buffer = self._write_buffer[len(data):]
+
+                if self._stanza_type == ibb_xso.IBBStanzaType.IQ:
+                    stanza = aioxmpp.IQ(
+                        aioxmpp.IQType.SET,
+                        to=self._peer_jid,
+                        payload=ibb_xso.Data(
+                            self._sid,
+                            self._outgoing_seq,
+                            data
+                        )
+                    )
+                elif self._stanza_type == ibb_xso.IBBStanzaType.MESSAGE:
+                    stanza = aioxmpp.Message(
+                        aioxmpp.MessageType.NORMAL,
+                        to=self._peer_jid,
+                    )
+                    stanza.xep0047_data = ibb_xso.Data(
+                        self._sid,
+                        self._outgoing_seq,
+                        data
+                    )
+
+                try:
+                    yield from self._service.client.send(
+                        stanza
+                    )
+                    self._outgoing_seq += 1
+                except errors.XMPPCancelError as e:
+                    self._connection_closed(e, cancel=False)
+                    return
+
+            if len(self._write_buffer) < self._output_buffer_limit_low:
+                if self._protocol is not None:
+                    self._protocol.resume_writing()
+
+            if not self._write_buffer:
+                if self._closing:
+                    break
+                self._can_write.clear()
+
+        close = ibb_xso.Close()
```

Comment:
I think you thought this through, but from a quick glance it’s not clear to me that this is sufficiently safe. Don’t we always want to send a Close, even if the write task fails with an exception? What is the suggested flow in the XEP here?

Reply:
Yes, you are right, the XEP mandates to close the stream when the packet data is erroneous.  My reading skills seem to have been temporarily disabled. This also means we have to keep sessions open, when we received erroneous data (and wait for the close). The only case when we do not send close is when the other side sent a close request or if the other side "vanished". This needs more careful handling of the possible errors.

Reply:
In case of error we can also just send a close and ignore the response, if it ever arrives. This can be achieved by using ``enqueue`` instead of ``send``.